rust-sdk/core/src/math/token.rs

```diff
@@ -0,0 +1,477 @@
+///
+/// # Returns
+/// - `u128`: The next square root price
+#[cfg_attr(feature = "wasm", wasm_bindgen(js_name = getNextSqrtPriceFromA, skip_jsdoc))]
+pub fn get_next_sqrt_price_from_a(
+    current_sqrt_price: U128,
+    current_liquidity: U128,
+    amount: U128,
+    specified_input: bool,
+) -> U128 {
+    if amount == 0 {
+        return current_sqrt_price;
+    }
+    let current_sqrt_price: u128 = current_sqrt_price.into();
+    let current_liquidity: u128 = current_liquidity.into();
+
+    let p = <U256>::from(current_sqrt_price).saturating_mul(amount.into());
+    let numerator = <U256>::from(current_liquidity)
+        .saturating_mul(current_sqrt_price.into())
+        .shl(64);
+
+    let current_liquidity_shifted: u128 = current_liquidity.shl(64);
+    let denominator = if specified_input {
+        current_liquidity_shifted + p
+    } else {
+        current_liquidity_shifted - p
+    };
+
+    let quotient: U256 = numerator / denominator;
+    let remainder: U256 = numerator % denominator;
+
+    let result: u128 = if remainder != 0 {
+        (quotient + 1).try_into().unwrap()
+    } else {
+        quotient.try_into().unwrap()
+    };
+
+    result.into()
+}
+
+/// Calculate the next square root price
+///
+/// # Parameters
+/// - `current_sqrt_price`: The current square root price
+/// - `current_liquidity`: The current liquidity
+/// - `amount`: The amount
+/// - `specified_input`: Whether the input is specified
+///
+/// # Returns
+/// - `u128`: The next square root price
+#[cfg_attr(feature = "wasm", wasm_bindgen(js_name = getNextSqrtPriceFromB, skip_jsdoc))]
+pub fn get_next_sqrt_price_from_b(
+    current_sqrt_price: U128,
+    current_liquidity: U128,
+    amount: U128,
+    specified_input: bool,
+) -> U128 {
+    if amount == 0 {
+        return current_sqrt_price;
+    }
+    let current_sqrt_price: u128 = current_sqrt_price.into();
+
+    let current_liquidity = <U256>::from(current_liquidity);
+    let amount_shifted = <U256>::from(amount).shl(64);
+
+    let quotient: U256 = amount_shifted / current_liquidity;
+    let remainder: U256 = amount_shifted % current_liquidity;
+
+    let delta: u128 = if !specified_input && remainder != 0 {
+        (quotient + 1).try_into().unwrap()
+    } else {
+        quotient.try_into().unwrap()
+    };
+
+    if specified_input {
+        (current_sqrt_price + delta).into()
+    } else {
+        (current_sqrt_price - delta).into()
+    }
+}
+
+/// Calculate the amount after transfer fee
+/// amount_without_transfer_fee > amount_with_transfer_fee
+///
+/// # Parameters
+/// - `amount`: The amount before tranfer fee
+/// - `transfer_fee`: The transfer fee
+/// - `adjust_up`: Whether to adjust up or down
+///
+/// # Returns
+/// - `u128`: The amount after transfer fee
+#[cfg_attr(feature = "wasm", wasm_bindgen(js_name = adjust_amount, skip_jsdoc))]
+pub fn adjust_amount(amount: U128, adjust_type: AdjustmentType, adjust_up: bool) -> U128 {
```

Comment:
I think for transfer fee we would need all 4 cases
* Swapping exactIn we know a we need to do `a with fee` -> `a without fee`. Then we get `b without fee` -> `b with fee`
* Swapping exactOut we do the reverse so `b with fee` -> `b without fee` and then `a without fee` -> `a with fee`

Reply:
@wjthieme 
At least, Whirlpool Program uses only 2 logic, so if we truely 4 logic, it is critical issue.

In ExactIn:
[ for input token ]
- user specify "Pre fee amount (transfer fee included amount)"
- the program calculate "Post fee amount (transfer fee exluced amount)"
- swap core logic use "Post fee amount" (swap core logic don't need to handle transfer fee)

[ for output token ]
- swap core logic generate output amount and it is "Pre fee amount"
- the program calculate "Post fee amount (transfer fee excluded amount)" and compare it with other_amount_threshold

User spacify amount transferred from their wallet as amount, and specify amount received as other_amount_threshold.

In ExactOut:
[ for output token ]
- user specify "Post fee amount (transfer fee excluded amount)"
- the program calculate "Pre fee amount"
- swap core logic use "Pre fee amount" (swap core logic don't need to handle transfer fee)

[ for input token ]
- swap core logic generate required input amount and it is "Post fee amount"
- the program calculate "Pre fee amount (transfer fee included amount)" and compare it with other_amount_threshold

Since the keyword adjust is somewhat ambiguous to me, I would like to see three separate functions (TransferFee, Slippage, SwapFee) for different purposes.
The fact that they call a common code internally would be a way to maintain readability.
However, if the common function branches for each AdjustmentType, it would be safer to separate the functions originally.

Reply:
Yeah I'll have a deeper look at this. You might be right. Your suggestion of implementing explicit functions (that all call the same common code) sounds like a good plan!